### PR TITLE
TST: Deprecate assert_numpy_array_equivalent

### LINKED ISSUE #10427: TST: Allow assert_numpy_array_equal to compare nan
Related to #10373.  ``pandas`` ``assert_array_equal`` needs to be updated to perform ``nan`` comparison properly.

```
import numpy as np
import pandas as pd
import pandas.util.testing as pdt
import numpy.testing.utils as npt

npt.assert_array_equal(np.array([1, np.nan, 3]), np.array([1, np.nan, 3]))
# no AssertionError

pdt.assert_numpy_array_equal(np.array([1, np.nan, 3]), np.array([1, np.nan, 3]))
# AssertionError: [  1.  nan   3.] is not equal to [  1.  nan   3.].
```

### PR DESCRIPTION
**Pull request by sinhrks**
Closes #10427.

### CODE REVIEW

**Comment by jreback** [comment]
add this to the deprecation issue for removal in 0.18

**Comment by sinhrks** [comment]
Sure. Also added ``testing.assert_isinstance`` #10458.

**Comment by jorisvandenbossche** [comment]
Is there a reason we don't use `np.testing.assert_array_equal` (but our own `array_equivalent` implementation)?

**Comment by jreback** [comment]
``array_equivalent`` is much more general as it handles all nulls for any dtypes (while numpy is restricted to floats). I may do nat's but I don't think so.

**Comment by jorisvandenbossche** [comment]
@sinhrks Can you then add a note in the docstring of `assert_numpy_array_equal` that it is similar to `np.testing.assert_array_equal` but more flexible regarding NaNs (so it is clear why this function exists/is used)

File: pandas/tests/test_index.py [hunk 1]
```diff
@@ -1554,15 +1553,14 @@ def test_equals_op(self):
         index_a = Index(['foo', 'bar', 'baz'])
         index_b = Index(['foo', 'bar', 'baz', 'qux'])
         # Testing Numpy Results Equivelent
-        assert_array_equal(
-            index_a.equals(index_a),
-            index_a == index_a
-        )
-        assert_array_equal(
+        self.assertTrue(index_a.equals(index_a))
```
Comment (sinhrks) [inline pandas/tests/test_index.py]: Existing test case looks incorrect. ``Index.equals`` returns ``bool`` when both index are equal.

```
pd.Index([1, 2, 3]).equals(pd.Index([1, 2, 3]))
# True
```

**Comment by sinhrks** [comment]
OK, also I found some test case uses ``numpy.testing.assert_array_equal`` imported to ``pandas.util.testing`` namespace. Fixed them and deprecate ``assert_array_equal`` also.

**Comment by sinhrks** [comment]
Ah,  ``np.testing.assert_array_equal`` behaves slightly different from current ``assert_numpy_array_equal``. I'm going to fix test cases to pass ``assert_numpy_array_equal`` which is more strict. Lmk if I should leave ``np.testing.assert_array_equal`.

```
np.testing.assert_array_equal(np.array([1, 1]), 1)
# no error 

pd.util.testing.assert_numpy_array_equal(np.array([1, 1]), 1)
# AssertionError: [1 1] is not equivalent to 1.
```

**Comment by sinhrks** [comment]
Found some cases are affected by  #10546. Revisit later.

**Comment by sinhrks** [comment]
Applied a fix done in #10546. I think this is ready for review.

**Comment by jorisvandenbossche** [comment]
@sinhrks looks good to me! (just needs a rebase)

Do we have a general rule for ourselves whether to use `tm.assert_...` or `self.assert_...`? They are available on both, but I saw you now changed some from self to tm.

File: pandas/util/testing.py [hunk 2]
```diff
@@ -669,6 +647,19 @@ def assert_numpy_array_equivalent(np_array, assert_equal, strict_nan=False, err_
         err_msg = '{0} is not equivalent to {1}.'.format(np_array, assert_equal)
     raise AssertionError(err_msg)
 
+def assert_numpy_array_equivalent(np_array, assert_equal,
```
Comment (jreback) [inline pandas/util/testing.py]: I don't think we need to deprecate this. Just remove it. These are internal testing routines that we don't want people to use.

**Comment by jreback** [comment]
the usual issue are we promoting using ``self.assert_numpy_array_equal`` or ``tm.assert_numpy_array_equal``. prob don't care so as long as both work.

pls rebase right before you merge as this touches lots of code.

File: pandas/util/testing.py (hunk 2 above)
Comment (sinhrks) [inline pandas/util/testing.py]: OK, will remove.

**Comment by sinhrks** [comment]
@jorisvandenbossche There is no policy, sometimes ``self``, ``tm.`` or directly import each funcs. 

Merging this.

**Comment by jorisvandenbossche** [comment]
Super!

**Comment by jreback** [comment]
very nice

now all we need is some kind of checker script that we can run in Travis to flag things 
kind of like a linter

eg might be as simple as greping for
np.testing and such

**Pull request merged**